pkg/template/processor.go

```diff
@@ -0,0 +1,111 @@
+package template
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/go-logr/logr"
+	templatev1 "github.com/openshift/api/template/v1"
+	"github.com/openshift/library-go/pkg/template/generator"
+	"github.com/openshift/library-go/pkg/template/templateprocessing"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/api/errors"
+	errs "github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"context"
+)
+
+type Processor struct {
+	cl     client.Client
+	scheme *runtime.Scheme
+	log    logr.Logger
+}
+
+func NewProcessor(cl client.Client, scheme *runtime.Scheme, log logr.Logger) *Processor {
+	return &Processor{cl: cl, scheme: scheme, log: log}
+}
+
+func (p Processor) ProcessAndApply(tmplContent []byte, values map[string]string) error {
+	objs, err := p.Process(tmplContent, values)
+	if err != nil {
+		return err
+	}
+
+	return apply(p.cl, objs, p.log)
+}
+
+func apply(cl client.Client, objs []runtime.RawExtension, rq logr.Logger) error {
+	for _, obj := range objs {
+		if obj.Object == nil {
+			rq.Info("template object is nil")
+			continue
+		}
+		gvk := obj.Object.GetObjectKind().GroupVersionKind()
+		rq.Info("processing object", "version", gvk.Version, "kind", gvk.Kind)
+		if err := cl.Create(context.TODO(), obj.Object); err != nil {
+			if errors.IsAlreadyExists(err) {
+				// If client failed to create all resources(few created, few remaining) in the first run, then to avoid
+				// resource already existing errors for later runs.(it's like oc apply, don't fail to exists)
+				continue
+			}
+			rq.Error(err, "unable to create resource", "type", gvk.Kind)
+			return errs.Wrapf(err, "unable to create resource of type %s", gvk.Kind)
+		}
+	}
+	return nil
+}
+
+// Process processes the template with the given content
+func (p Processor) Process(tmplContent []byte, values map[string]string) ([]runtime.RawExtension, error) {
+	scheme := p.scheme
+	codecs := serializer.NewCodecFactory(scheme)
+	decode := codecs.UniversalDeserializer().Decode
+	obj, _, err := decode(tmplContent, nil, nil)
+	if err != nil {
+		p.log.Error(err, "unable to decode template")
+		return nil, err
+	}
+	tmpl := obj.(*templatev1.Template)
+	if err := injectUserVars(tmpl, values, false); err != nil {
+		p.log.Error(err, "unable to inject vars in template", "error", err.Error())
+		return nil, err
+	}
+	tmpl, err = processTemplateLocally(tmpl, scheme)
+	if err != nil {
+		p.log.Error(err, "unable to process template")
+		return nil, err
+	}
+	return tmpl.Objects, nil
+}
+
+// injectUserVars injects user specified variables into the Template
+func injectUserVars(t *templatev1.Template, values map[string]string, ignoreUnknownParameters bool) error {
+	for param, val := range values {
+		v := templateprocessing.GetParameterByName(t, param)
+		if v != nil {
+			v.Value = val
+			v.Generate = ""
+		} else if !ignoreUnknownParameters {
+			return fmt.Errorf("unknown parameter name %q", param)
```

Comment:
why would you fail if the template doesn't have the parameter? consider that we have 10 pre-configured variables and the templates use only subsets of them. There is no point in failing
In addition, consider custom templates - it shouldn't rely on the number of vars

Reply:
I agree. But logging a warning could still be useful for debugging.

Reply:
We will have multiple templates (even for one user there will be multiple templates for different namespaces). We don't really know what template uses what parameters. If we require the exact match between the passing parameters and the template then the controller will have to prepare the parameters individually for each template which can be quite tricky.
It could just generate the same parameters every time and pass them to the processor for all the templates instead.

Reply:
ok, now got it. will change this to not fail and log it as warning

Reply:
I guess that this is the reason why the tenant service didn't use this process logic, but implemented it's own replace - to have the flexibility.